downloader/src/main/kotlin/vcs/GitWorkingTree.kt

```diff
@@ -21,68 +21,85 @@ package com.here.ort.downloader.vcs
+    private val repo = FileRepositoryBuilder().readEnvironment().findGitDir(workingDir).runCatching {
+        build()
+    }.getOrNull()
+
+    override fun isValid(): Boolean = repo?.objectDatabase?.exists() == true
+
+    override fun isShallow(): Boolean = repo?.directory?.resolve("shallow")?.isFile == true
+
+    private fun listSubmodulePaths(repo: Repository): List<String> {
+        fun listSubmodules(parent: String, repo: Repository, paths: MutableList<String>) {
+            val prefix = if (parent.isEmpty()) parent else "$parent/"
+
+            SubmoduleWalk.forIndex(repo).use { walk ->
+                while (walk.next()) {
+                    paths += "$prefix${walk.path}"
+                    walk.repository.use { submoduleRepo ->
+                        listSubmodules(paths.last(), submoduleRepo, paths)
+                    }
+                }
+            }
         }
 
-        // Do not use runGitCommand() here as we do not require the command to succeed.
-        val isInsideWorkTree = ProcessCapture(workingDir, "git", "rev-parse", "--is-inside-work-tree")
-        return isInsideWorkTree.isSuccess && isInsideWorkTree.stdout.trimEnd().toBoolean()
+        return mutableListOf<String>().also { paths ->
+            listSubmodules("", repo, paths)
+        }
     }
 
-    override fun isShallow(): Boolean {
-        val dotGitDir = gitBase.run(workingDir, "rev-parse", "--absolute-git-dir").stdout.trimEnd()
-        return File(dotGitDir, "shallow").isFile
-    }
+    override fun getNested(): Map<String, VcsInfo> =
+        runCatching {
+            // This throws a NullPointerException if "repo" is null.
+            val paths = listSubmodulePaths(repo!!)
+            paths.associateWith { GitWorkingTree(repo.workTree.resolve(it), gitBase).getInfo() }
+        }.getOrElse {
+            emptyMap()
+        }
 
-    override fun getNested(): Map<String, VcsInfo> {
-        val root = getRootPath()
-        val paths = gitBase.run(root, "submodule", "--quiet", "foreach", "--recursive", "echo \$displaypath").stdout
-            .lines().filter { it.isNotBlank() }
+    override fun getRemoteUrl(): String =
+        runCatching {
+            // This throws a NullPointerException if "repo" is null.
+            val remotes = org.eclipse.jgit.api.Git(repo!!).remoteList().call()
+            val remoteForCurrentBranch = BranchConfig(repo.config, repo.branch).remote
 
-        return paths.associateWith { GitWorkingTree(root.resolve(it), gitBase).getInfo() }
-    }
+            val remote = if (remotes.size <= 1 || remoteForCurrentBranch == null) {
+                remotes.firstOrNull()
+            } else {
+                remotes.find { remote ->
+                    remote.name == remoteForCurrentBranch
+                }
+            }
 
-    override fun getRemoteUrl() =
-        gitBase.run(workingDir, "remote", "get-url", getRemoteForCheckout()).stdout.trimEnd()
+            remote?.urIs?.firstOrNull()?.toString()
```

Comment:
I guess `urIs` is short for `getURIs()`, but do you know why it isn't `uRIs`?

Reply:
Interesting, I didn't even notice this thanks to auto-completion. Looks like a Java-to-Kotlin name mapping bug to my, similar to [this one](https://youtrack.jetbrains.com/issue/KT-17768).

Reply:
Interesting, unfortunately it doesn't seem that they will fix it.